src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -1300,6 +1333,70 @@ private static ErrorValue GetDiv0Error(IRContext irContext)
             });
         }
 
+        // Functions that return an integer result, such as Len, will return a Decimal by default or
+        // a Float if compiled under NumberIsFloat.  Decimal is preferrable for decimal centric calculations
+        // as a Float will promote the entire expression to Float and possibly lose precision.
+        //
+        // Could these variations be done as an overload?  Yes. But we really want to make sure we get the
+        // right onw and not have C# coerce to another data type and possibly lose precision.
+        private static FormulaValue NumberOrDecimalValue(IRContext irContext, int value)
+        {
+            // all int values fit in both double and decimal
+            if (irContext.ResultType == FormulaType.Number)
+            {
+                return new NumberValue(irContext, (double)value);
+            }
+            else if (irContext.ResultType == FormulaType.Decimal)
+            {
+                return new DecimalValue(irContext, (decimal)value);
+            }
+            else
+            {
+                return CommonErrors.UnreachableCodeError(irContext);
+            }

```

Comment:
(nit) I would make it more readable with this 

return irContext.ResultType._type.Kind switch
            {
                DKind.Number => new NumberValue(irContext, (double)value),
                DKind.Decimal => new DecimalValue(irContext, (decimal)value),
                _ => CommonErrors.UnreachableCodeError(irContext)
            };

Reply:
I think I'll leave it for now.  I don't like reaching into the _type if not needed, and that patterns isn't used in LibraryMath.cs.  It is used in LibraryData.cs however.  :)  Something we should apply consistently, but that can happen at another time.